### ISSUE TITLE
Reducing memory usage during image load

### ISSUE THREAD

**Issue by rgov** [manoharan-lab/holopy#267]
In troubleshooting high memory usage of a process running HoloPy, I noticed that loading a grayscale TIF file takes approximately 200% more memory than necessary, and the memory usage does not subside quickly.

1. `load_image` calls `data_grid` and passes in the image data, which is say ~120 MB of grayscale values represented as doubles.

2. `data_grid` executes `arr=np.array([arr])` which copies the array.

3. `data_grid` later calls `update_metadata` which directly calls `a.copy()` and returns the copy.

Theoretically these shouldn't consume memory over time because of Python's reference counting, but I do not see the memory subsiding.

For (2), I found that `arr = np.expand_dims(arr, axis=0)` produces the same array shape without the overhead.

For (3) the copy is probably not necessary at load time and could be skipped.

---

Additionally, during a call to `load_average`, there is a large spike in memory usage. At the peak of computation it requires 600 MB of RAM *per* input TIF file (4864 × 3232 pixels, grayscale).

1. I believe `clean_concat` copies all of the image data into a single array.

2. `accumulator.mean('images')` causes a temporary spike that partially subsides. Note that this calculation can be saved and re-used instead of performed twice. The resulting object occupies about the same size as a single image.

3. The calculation of `noise_sd` also causes a spike that partially subsides.

4. `copy_metadata` (optionally) and then `update_metadata` both make copies of their inputs.

**Comment by barkls** [comment]
Thank you for finding these.

For your first set of comments on load_image, I agree your solution to point 2 is better. There might be some edge cases where the copy() in point 3 is necessary - I'll take a look.

For your second set of comments on load_average, this is an issue that is actually already on our radar, but not posted on github. @ralex0 are you looking into this?

**Comment by ralex0** [comment]
Yeah, I am actually already working on the load_average performance. @briandleahy noticed this same issue before. 

@rgov, what branch are you using? Your clone of the master?

**Comment by rgov** [comment]
I am using HoloPy installed from Conda Forge. I don’t know the version at the moment.

**Comment by rgov** [comment]
As I understand it:

* `copy_metadata` returns a copy of an input array, attaching metadata copied from a second input array (potentially the same one)

* `update_metadata` returns a copy of an input array, replacing certain metadata fields with the given input parameters.

In many cases in the code, I see patterns like this that result in unnecessary copies:

```python
# makes a copy of data and then releases the original, so that only the copy remains
data = update_metadata(data, medium_index=medium_index, illum_wavelen=illum_wavelen)

# creates an array with detector_grid, copies it, and then releases the original
data = update_metadata(detector_grid(shape=10, spacing=2), 1.33, 0.660, (0,1))
```

Since copying is often not required, what do you think about changing the semantics of these routines so that the developer is more explicit about copies?

* Change `update_metadata` to mutate its parameter, consistent with Python's `dict.update()` and `set.update()`. Return the input array.
* Add a `from` parameter to `update_metadata` that specifies another array to copy metadata from. Copy metadata from this array first, then override any additional arguments that are passed in.
* Remove `copy_metadata`.
* Add an explicit `copy_array(src)` function, if one does not already exist, which also copies the metadata from `src`.

The examples above would look like:

```
update_metadata(data, medium_index=medium_index, illum_wavelen=illum_wavelen)
data = update_metadata(detector_grid(shape=10, spacing=2), 1.33, 0.660, (0,1))
```

When a copy is needed, it would be written like this:

```python
# before:
yschema = update_metadata(xschema, illum_polarization=ypolarization)

# after:
yschema = update_metadata(copy_array(xschema), illum_polarization=ypolarization)
```

This would be a breaking change so if these are APIs third-parties are relying on, you would need to have a plan to handle the transition.

**Comment by rgov** [comment]
Alternatively, somewhat more consistent with Numpy, you could add an optional `out` parameter that specifies the array to write to. If it's not provided, a copy is made.

**Comment by barkls** [comment]
Fixed by #276 .
